ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -15,52 +15,256 @@
-import org.apache.kafka.common.utils.Time;
-import org.apache.kafka.streams.KafkaStreams;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.streams.StreamsConfig;
 
-public class UtilizationMetricsListener implements Runnable, QueryEventListener {
+public class UtilizationMetricsListener implements MetricsReporter {
+  private final Map<String, List<TaskStorageMetric>> metricsSeen;
+  private final Metrics metricRegistry;
+  private static final Object lock = new Object();
+  private static List<MetricName> registeredNodeMetrics = null;
 
-  private final List<KafkaStreams> kafkaStreams;
-  private final Logger logger = LoggerFactory.getLogger(UtilizationMetricsListener.class);
-  private final List<String> metrics;
-  private final Time time;
-  private long lastSampleTime;
 
   public UtilizationMetricsListener() {
-    this.kafkaStreams = new ArrayList<>();
-    this.metrics = new LinkedList<>();
-    time = Time.SYSTEM;
-    lastSampleTime = time.milliseconds();
+    this(MetricCollectors.getMetrics());
+  }
+
+  public UtilizationMetricsListener(final Metrics metricRegistry) {
+    this.metricsSeen = new HashMap<>();
+    this.metricRegistry = metricRegistry;
+  }
+
+  // for testing
+  public UtilizationMetricsListener(
+      final Metrics metricRegistry,
+      final Map<String, List<TaskStorageMetric>> taskMetrics) {
+    this.metricsSeen = taskMetrics;
+    this.metricRegistry = metricRegistry;
   }
 
   @Override
-  public void onCreate(
-      final ServiceContext serviceContext,
-      final MetaStore metaStore,
-      final QueryMetadata queryMetadata) {
-    kafkaStreams.add(queryMetadata.getKafkaStreams());
+  public void init(List<KafkaMetric> list) {
+
   }
 
+
   @Override
-  public void onDeregister(final QueryMetadata query) {
-    final KafkaStreams streams = query.getKafkaStreams();
-    kafkaStreams.remove(streams);
+  public void configure(Map<String, ?> map) {
+    String dir;
+    if (map.containsKey(StreamsConfig.STATE_DIR_CONFIG)) {
+      dir = map.get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    } else {
+      dir =  StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    }
+    configureShared(new File(dir), this.metricRegistry);
+  }
+
+  private static void configureShared(final File baseDir, final Metrics metricRegistry) {
+    synchronized (lock) {
+      if (registeredNodeMetrics != null) {
```

Comment:
Rather than keeping the registered metrics in a collection, why not just use a single boolean that you set to true once the node-level metrics are registered? You actually don't even need a lock in that case - it can just be an atomic boolean:

```
static AtomicBoolean  registeredNodeMetrics = new AtomicBoolean(false);

private static void configureShared(final File baseDir, final Metrics metricsRegistry) {
    if (registeredNodeMetrics.getAndSet(true)) {
        // already registered;
        return;
    }
    // add the node level metrics
}
```